### ISSUE TITLE
object* obs = objects(); foreach(object ob in obs) printf("%O: %d\n", ob, refs(ob)); 引起宕机

### ISSUE THREAD

**Issue by suimengjushi** [fluffos/fluffos#750]
```
Error in mudlib error handler: *Undefined function called: query_temp
program: /feature/name.c, object: /feature/name, file: /feature/name.c:46
--- trace ---
Object: /clone/user/user#8, Program: /feature/command.c
   in command_hook() at /feature/command.c:43
arguments: [0]
locals: ["cpux", "/cmds/adm/cpux", 0, /d/city3/chatroom1]
Object: /cmds/adm/cpux, Program: /cmds/adm/cpux.c
   in main() at /cmds/adm/cpux.c:5
arguments: [/clone/user/user#8 ("欢迎")]
locals: [Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
({ /* sizeof() == 302 */
  /cmds/adm/cpux,
  /clone/weapon/changjian#106 ("长剑"),
  /clone/weapon/changjian ("长剑"),
  /clone/cloth/cloth#105 ("布衣"),
  /cmds/app/clone,
  /d/wizard/fightroom5,
  /d/city3/chatroom1,
  /clone/money/silver#104 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#103 ("布衣"),
  /d/city3/npc/chike#102 ("吃客"),
  /clone/money/silver#101 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#100 ("布衣"),
  /d/city3/npc/chike ("吃客"),
  /clone/cloth/cloth#99 ("布衣"),
  /d/city3/npc/flowerg#98 ("卖花姑娘"),
  /clone/cloth/cloth#97 ("布衣"),
  /d/city3/npc/flowerg ("卖花姑娘"),
  /d/city3/xijie2,
  /d/city3/dongjie1,
  /d/city3/beijie2,
  /d/city3/nanjie1,
  /clone/armor/tiejia#96 ("铁甲"),
  /clone/weapon/gangjian#95 ("钢剑"),
  /d/npc/shandian#94 ("[1;37m闪电[2;37;0m"),
  /clone/armor/tiejia#93 ("铁甲"),
  /clone/weapon/gangjian#92 ("钢剑"),
  /kungfu/skill/shenxing-baibian,
  /d/npc/shandian ("[1;37m闪电[2;37;0m"),
  /clone/money/silver#91 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#90 ("布衣"),
  /d/city3/npc/laolanyan#89 ("老烂眼"),
  /clone/money/silver#88 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#87 ("布衣"),
  /d/city3/npc/laolanyan ("老烂眼"),
  /clone/money/silver#86 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#85 ("布衣"),
  /d/city3/npc/xiaolanyan#84 ("小烂眼"),
  /clone/money/silver#83 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#82 ("布衣"),
  /d/city3/npc/xiaolanyan#81 ("小烂眼"),
  /clone/money/silver#80 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#79 ("布衣"),
  /d/city3/npc/xiaolanyan#78 ("小烂眼"),
  /clone/money/silver#77 ("[37m银子[2;37;0m"),
  /clone/cloth/cloth#76 ("布衣"),
  /d/city3/npc/xiaolanyan ("小烂眼"),
  /d/city3/guangchang,
  /kungfu/condition/drunk,
  /clone/cloth/cloth#75 ("布衣"),
  /d/city/npc/xiaoer2#74 ("店小二"),
  /clone/cloth/cloth#73 ("布衣"),
  /d/city/npc/xiaoer2 ("店小二"),
   .
   .
   .省略一堆
   .
   .
  /feature/save,
  /feature/dbase,
  /feature/treemap,
  /adm/daemons/virtuald,
  /adm/daemons/securityd,
  /adm/obj/master,
  /adm/obj/simul_efun
}), Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
/feature/name]
Object: /adm/obj/simul_efun, Program: /adm/obj/simul_efun.c
   in printf() at /adm/simul_efun/message.c:380
arguments: ["%O: %d
", Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
/feature/name, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
Object: /adm/obj/master, Program: /adm/obj/master.c
   in object_name() at /adm/obj/master.c:437
arguments: [Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
/feature/name]
Object: /feature/name, Program: /feature/name.c
   in name() at /feature/name.c:46
arguments: [0]
locals: [0, 0]
Object: /adm/obj/master, Program: /adm/obj/master.c
   in error_handler() at /adm/obj/master.c:321
arguments: [Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.
Error '*Undefined function called: query_temp
' while trying to print error trace -- trace suppressed.

name.c 

varargs string name(int raw)
{
	string str, *mask;

	if( !raw && sizeof(mask = query_temp("apply/name")) )  //这行报错
		return mask[sizeof(mask)-1];
	else
	{
		if( stringp(str = query("name")) )
			return str;
		else
			return file_name(this_object());
	}
}

以下北大 替换就OK
varargs string name(int raw)
{
	string str, *mask;
	int size = 0;
	object link_ob;
	//在name()里调用不存在的函数会crash.一般name.c都会和dbase.c同时被继承
	//所以没有问题，但是如果直接call name.c的函数，这里就会出错
	//注意，即使你call name.c->otherfoo()，mudos也会主动调用name()

	if(function_exists("query",this_object(),0) == 0)
	{
		return file_name(this_object());
	}
	if( !raw )
	{
		mask = query_temp("apply/name");
		if(!undefinedp(mask)) size = sizeof(mask);
		if(size > 0)
			return mask[size-1];
	}
	if( stringp(str = query("name")) )
	{
		if(userp(this_object()) && this_object()->is_realuser())
		{
    		return HIY"Ｖ"NOR+str;
		}
		return str;
	}
	else
		return file_name(this_object());
}
```

**Comment by XMLSDK** [comment]
你有否加回lonely的PACKAGE_DATABASE ?
https://github.com/fluffos/fluffos/issues/341
https://github.com/lonely21/fluffos/tree/master/src/packages/database

**Comment by suimengjushi** [comment]
并没有，因为本身用的不是泥潭的lib

**Comment by XMLSDK** [comment]
你加回lonely的PACKAGE_DATABASE才能用query_temp

**Comment by seikichin** [comment]
所答非所问，他说的和你说的没关系

**Comment by suimengjushi** [comment]
是的 ，群里有个大佬的解释应该是比较合理的

从报错信息来看，name.c在内存中也是以一个object存在的，而这个object没有继承dbase，所以当执行到 name.c->name()的时候，query和query_temp都会出错，并且name.c->name()会反复调用自己，导致死循环？

**Comment by thefallentree** [comment]
in handling error, the driver tries to dump a trace, which internally calls into VM again through (sprintf svalue_to_string() ).  This is going to be fixed somewhat in the new release

**Comment by thefallentree** [comment]
this is now fixed in #735 with a test
